vllm/adapter_commons/models.py

```diff
@@ -0,0 +1,108 @@
+
+from torch import nn
+
+from vllm.logger import init_logger
+from vllm.utils import LRUCache
+
+logger = init_logger(__name__)
+
+
+class AdapterModel:
+
+    def __init__(self, model_id=None):
+        self.id = model_id
+
+    @classmethod
+    def from_local_checkpoint(cls, model_dir, model_id=None, **kwargs):
+        # Common initialization code
+        # Load weights or embeddings from local checkpoint
+        raise NotImplementedError("Subclasses must implement this method.")
+
+
+T = TypeVar('T')
+
+
+class AdapterLRUCache(LRUCache[T]):
+
+    def __init__(self, capacity: int, deactivate_fn: Callable[[Hashable],
+                                                              None]):
+        super().__init__(capacity)
+        self.deactivate_fn = deactivate_fn
+
+    def _on_remove(self, key: Hashable, value: T):
+        logger.debug("Removing adapter int id: %d", key)
+        self.deactivate_fn(key)
+        return super()._on_remove(key, value)
+
+
+class AdapterModelManager:
+
+    def __init__(
+        self,
+        model: nn.Module,
+    ):
+        """Create a AdapterModelManager and adapter for a given model.
+        Args:
+            model: the model to be adapted.
+        """
+        self.model: nn.Module = model
+        self._registered_adapters: Dict[int, Any] = {}
+        # Dict instead of a Set for compatibility with LRUCache.
+        self._active_adapters: Dict[int, None] = {}
+        self.adapter_type = 'Adapter'
+        self._last_mapping = None
+
+    def __len__(self) -> int:
+        return len(self._registered_adapters)
+
+    @abstractproperty
+    def adapter_slots(self):
+        ...
+
+    @abstractproperty
+    def capacity(self):
+        ...
+
+    def _deactivate_adapter(self, adapter_id: int):
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def deactivate_adapter(self, adapter_id: int) -> bool:
+        if adapter_id in self._active_adapters:
+            self._deactivate_adapter(adapter_id)
+            self._active_adapters.pop(adapter_id)
+            return True
+        return False
+
+    def activate_adapter(self, adapter_id: int) -> bool:
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def _add_adapter(self, adapter: Any):
+        raise NotImplementedError("Subclasses must implement this method.")
+
+    def add_adapter(self, adapter: Any) -> bool:
+        if adapter.id not in self._registered_adapters:
+            if len(self._registered_adapters) >= self.capacity:
+                raise RuntimeError("No free " + self.adapter_type + " slots.")
+            self._add_adapter(adapter)
+            return True
+        return False
+
+    def set_adapter_mapping(self, mapping: Any) -> None:
+        if self._last_mapping != mapping:
+            self._set_adapter_mapping(mapping)
+        self._last_mapping = mapping
+
+    def _set_adapter_mapping(self, mapping: Any) -> None:
```

Comment:
@Yard1  the reason I implemented this private method is because `set_adapter_mapping` uses this internal method which needs to be implemented on the base adapter class level. 
If I only set this method in the child classes like _set_lora_mapping and _set_prompt_adapter mapping, then I won't be able abstract the common code in `set_adapter_mapping`.
Is there another way to do this?

Reply:
We may need to just redesign the API (which is fine).